### ISSUE TITLE
[WAGON-564] SSH connection failure because 'preferredAuthentications' option is ignored if password isn't set

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-wagon#624]
**[Lucian Burja](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lburja)** opened **[WAGON-564](https://issues.apache.org/jira/browse/WAGON-564?redirect=false)** and commented

I am trying to upload a file via SSH with private key authentication, using the `wagon-maven-plugin` plugin. The Linux server that is also integrated with Kerberos (which I don't use).

Although I provide a valid `privateKey`, and I set `<preferredAuthentications>publickey</preferredAuthentications>,` the Kerberos authentication is always triggered.

While investigating, I have found the following root cause:

In settings.xml, for a `<server>` you can decide to use SSH key based authentication instead of username/password:

```java
<server>
   <id>myserver</id>
   <username>bamboo</username>
   <privateKey>...path to the file...</privateKey>

   <configuration>
      <preferredAuthentications>publickey</preferredAuthentications>      
   </configuration>
</server>
```

According to the documentation, this authentication option only works if you omit the `password` element, otherwise `privateKey` is ignored.

However, if `password` is omitted, then `preferredAuthentications` is ignored, as can be seen in `AbstractJschWagon.java :: openConnectionInternal  (line 254)`

```java
if ( authenticationInfo.getPassword() != null )
{
    config.setProperty( "PreferredAuthentications", preferredAuthentications );
}
```

 

Thus, in practice, if you use `privateKey` based authentication, you cannot control the `PreferredAuthentications` parameter, and the default value is used: `gssapi-with-mic,publickey,password,keyboard-interactive`. This triggers Kerberos based authentication as the first option.

A simple patch to solve this issue is to add to the lines above an else branch, like this:

```java
if ( authenticationInfo.getPassword() != null )
{
    config.setProperty( "PreferredAuthentications", preferredAuthentications );
}
else if ( !"gssapi-with-mic,publickey,password,keyboard-interactive".equals( preferredAuthentications ) )
{
    // if different then the default, always set
    config.setProperty( "PreferredAuthentications", preferredAuthentications );
}
```

 or to remove the the surrounding if-statement all-together 

---

**Affects:** 3.3.3

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Looking into the code. This looks like non-sense to me. There is always a default value you can override. Can you try a wagon snapshot? I would provide a patch. The cause is WAGON-237.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

According to the [change log](http://www.jcraft.com/jsch/ChangeLog) we have:

> Changes since version 0.1.30:
> - feature: an option "PreferredAuthentications" is added.
>   The default value is "gssapi-with-mic,publickey,keyboard-interactive,password".

So, I'd remove the default value and will set your custom value if it is not empty. This should solve your issue. WDYT?

**Comment by jira-importer** [comment]
**[Lucian Burja](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lburja)** commented

Yes, if you provide a patch, or a git branch I can try my particular case!

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Branch pushed.

**Comment by jira-importer** [comment]
**[Lucian Burja](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lburja)** commented

I've tested the branch WAGON-564, and it works as expected:
* the value in preferredAuthentications is correctly taken into account for privateKey authentication
* when the preferredAuthentications element is not present, the default configuration applies

Thank you! Looking forward for the 3.3.4 release!

As a side note, it would be great to also mention these configuration options on the website; currently it's very hard to know that these options even exist.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will merge the branch shortly.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [184cdad07a78cedcfdce68b6d8b05040593f4176](https://gitbox.apache.org/repos/asf?p=maven-wagon.git;a=commit;h=184cdad07a78cedcfdce68b6d8b05040593f4176).
